website/versioned_docs/version-v1.x/circuits.md

```diff
@@ -7,45 +7,407 @@ sidebar_position: 7
+
+### Circom
+
+MACI's circuits are written using [Circom](https://docs.circom.io/), a domain-specific language (DSL) used to write zk-SNARK circuits. Circom syntax resembles JavaScript, and it currently is one of the most popular DSL in use by zk developers. Please refer to their documentation to learn more about the language.
+
+### Proving system
+
+MACI uses [GROTH16](https://eprint.iacr.org/2016/260.pdf) as its proving system. GROTH16 is a zk-SNARK proving system that allows for the generation of proofs that are small and fast to verify.
+
+## How are the circuits used
+
+The circuits are used by the coordinator (the prover) to prove that they have correctly processed a batch of messages and tallied the votes correctly. This happens after a Poll has completed, and the coordinator has merged the state and message trees. The coordinator then generates a proof for each batch of messages, and submits them to the contract. The contract then verifies the proofs and updates the commitments on chain.
+
+> Please note that in order to function, the circuit in charge of processing the messages requires the coordinator's private key, hence a proof for this circuit can only be generated by the coordinator. The private key is needed in order to generate the ECDH shared key used to decrypt the messages.
+
+## How do the Circuits fit in a voting round
+
+![flow](/img/circuits/processingAfterPollEnds.svg)
+
+## How the circuits work
+
+### Message processing (processMessages)
+
+This circuit allows the coordinator to prove that they have correctly processed each message in reverse order, in a consecutive batch of 5 ^ msgBatchDepth messages to the respective state leaf within the state tree. Coordinators would use this circuit to prove correct execution at the end of each Poll. The pre-requisites for this are:
+
+- the related Poll has ended
+- the state tree has been merged
+- the message tree has been merged
+
+![ProcessMessages](/img/circuits/processMessages.svg)
+
+#### Parameters
 
 | #   | Parameter                | Description                                          |
 | --- | ------------------------ | ---------------------------------------------------- |
-| 0   | State tree depth         | Should be set to 10. Allows 9,765,625 signups.       |
+| 0   | State tree depth         | Allows $(5^{n})$ signups.                            |
 | 1   | Message tree depth       | Allows $(5^{n})$ votes or key-change messages.       |
 | 2   | Message batch tree depth | Allows $(5^{n})$ messages to be processed per batch. |
 | 3   | Vote option tree depth   | Allows $(5^{n})$ vote options.                       |
 
-### Vote tallying
+#### Inputs
+
+| Input signal                     | Description                                                                             |
```

Comment:
This table is great, thanks for doing this

Reply:
oh no can't take props for that, it's a copy pasta from here: https://maci.pse.dev/docs/spec 🙂